p2p/host.go

```diff
@@ -4,250 +4,228 @@ import (
-// message for sending.
-func (host *HostV2) SendMessageToGroups(groups []nodeconfig.GroupID, msg []byte) (err error) {
+func (h *hmyHost) SendMessageToGroups(groups []nodeconfig.GroupID, msg []byte) error {
+	ctx := context.Background()
+	var g errgroup.Group
+
 	for _, group := range groups {
-		t, e := host.getTopic(string(group))
-		if e != nil {
-			err = e
-			continue
-		}
-		e = t.Publish(context.Background(), msg)
-		if e != nil {
-			err = e
-			continue
+		top := string(group)
+		_, err := h.getTopic(top)
+		if err != nil {
+			return err
 		}
+		g.Go(func() error {
+			return h.coreAPI.PubSub().Publish(ctx, top, msg)
+		})
 	}
-	return err
+
+	return g.Wait()
 }
 
-// AddPeer add p2p.Peer into Peerstore
-func (host *HostV2) AddPeer(p *Peer) error {
-	if p.PeerID != "" && len(p.Addrs) != 0 {
-		host.Peerstore().AddAddrs(p.PeerID, p.Addrs, libp2p_peerstore.PermanentAddrTTL)
-		return nil
+func (h *hmyHost) getTopic(topic string) (ipfs_interface.PubSubSubscription, error) {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+	if t, ok := h.joined[topic]; ok {
+		return t, nil
 	}
 
-	if p.PeerID == "" {
-		host.logger.Error().Msg("AddPeer PeerID is EMPTY")
-		return fmt.Errorf("AddPeer error: peerID is empty")
-	}
+	sub, err := h.coreAPI.PubSub().Subscribe(context.Background(), topic)
 
-	// reconstruct the multiaddress based on ip/port
-	// PeerID has to be known for the ip/port
-	addr := fmt.Sprintf("/ip4/%s/tcp/%s", p.IP, p.Port)
-	targetAddr, err := ma.NewMultiaddr(addr)
 	if err != nil {
-		host.logger.Error().Err(err).Msg("AddPeer NewMultiaddr error")
-		return err
+		return nil, errors.Wrapf(err, "cannot join pubsub topic %x", topic)
 	}
 
-	p.Addrs = append(p.Addrs, targetAddr)
-	host.Peerstore().AddAddrs(p.PeerID, p.Addrs, libp2p_peerstore.PermanentAddrTTL)
-	host.logger.Info().Interface("peer", *p).Msg("AddPeer add to libp2p_peerstore")
-	return nil
+	h.joined[topic] = sub
+	return sub, nil
 }
 
-// Peerstore returns the peer store
-func (host *HostV2) Peerstore() libp2p_peerstore.Peerstore {
-	return host.h.Peerstore()
+type NamedSub struct {
+	Topic string
+	Sub   ipfs_interface.PubSubSubscription
 }
 
-// GetID returns ID.Pretty
-func (host *HostV2) GetID() libp2p_peer.ID {
-	return host.h.ID()
+func (h *hmyHost) AllSubscriptions() []NamedSub {
+	subs := []NamedSub{}
+	h.lock.Lock()
+	defer h.lock.Unlock()
+	for name, g := range h.joined {
+		subs = append(subs, NamedSub{name, g})
+	}
+	return subs
 }
 
-// GetSelfPeer gets self peer
-func (host *HostV2) GetSelfPeer() Peer {
-	return host.self
+func (h *hmyHost) GetPeerCount() int {
+	conns, _ := h.coreAPI.Swarm().Peers(context.Background())
+	return len(conns)
 }
 
-// GetP2PHost returns the p2p.Host
-func (host *HostV2) GetP2PHost() libp2p_host.Host {
-	return host.h
-}
+// NewHost ..
+func NewHost(opts *Opts, own *Peer) (Host, error) {
+	var swarmAddresses []string
+
+	if opts.Port != 0 {
+		swarmAddresses = []string{
+			fmt.Sprintf("/ip4/127.0.0.1/tcp/%d", opts.Port),
```

Comment:
these hard coded value needs updates?

Reply:
sort of - I don't think strictly necessray but we do allow clinet to put in IP so will try - (although I think if we deploy on stn, it will work as it)